gno.land/pkg/gnoland/app.go

```diff
@@ -194,9 +215,13 @@ func InitChainer(baseApp *sdk.BaseApp, acctKpr auth.AccountKeeperI, bankKpr bank
 	}
 }
 
-// XXX not used yet.
-func EndBlocker(vmk vm.VMKeeperI) func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-	return func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-		return abci.ResponseEndBlock{}
+// EndBlocker defines the logic executed after every block.
+// Currently, it parses events that happened during execution to calculate
+// validator set changes
+func EndBlocker(collector *collector[abci.ValidatorUpdate]) func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
+	return func(_ sdk.Context, _ abci.RequestEndBlock) abci.ResponseEndBlock {
+		return abci.ResponseEndBlock{
+			ValidatorUpdates: collector.getEvents(),
```

Comment:
looks good.

Reply:
I was on the fence about what we should shove in the `EndBlocker` logic, because ideally it should be super quick -- this is why I chose the approach of just having the changes being prepared in advance, and this call just triggering a lookup